### ISSUE TITLE
Rainbow Memory: Continual Learning with a Memory of Diverse Samples

### ISSUE THREAD

**Issue by chaerinkong** [reyllama/MIPAL-CL#1]
## TL; DR

- Keep diversity-aware memory through stratified sampling aligned by model prediction uncertainty.
- Incorporate powerful augmentations to further boost diversity.

## Motivation

- The memory used for rehearsal in continual learning should have be both representative (near center) and discriminative (near boundary)

## Formulation

![image](https://user-images.githubusercontent.com/46308407/158128214-2f4bddeb-5a7b-4229-99ca-652c1d5e0eea.png)

![image](https://user-images.githubusercontent.com/46308407/158128247-001c0d21-74c2-4043-9dda-641f35b8c53a.png)

- Compute uncertainty by observing model prediction after diverse perturbations (augmentation)

![image](https://user-images.githubusercontent.com/46308407/158128376-0b39c14c-297d-4da9-a7e6-01bb482fad71.png)

## Evaluation

![image](https://user-images.githubusercontent.com/46308407/158128416-a37c0c98-e40e-4109-a4c8-f6a635eaa4f3.png)
![image](https://user-images.githubusercontent.com/46308407/158128441-2bbacfab-75c9-4bc9-87ac-8fe167f6f548.png)
![image](https://user-images.githubusercontent.com/46308407/158128460-4682e5b9-bdd1-471f-9cd7-0c14dbfb02b5.png)
